server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -176,6 +176,18 @@
     private final KeyedLock<Long> noOpKeyedLock = new KeyedLock<>();
     private final AtomicBoolean shouldPeriodicallyFlushAfterBigMerge = new AtomicBoolean(false);
 
+    /**
+     * If multiple writes passed {@link InternalEngine#tryReserveDocs(Operation, int)} but they haven't adjusted
+     * {@link IndexWriter#getPendingNumDocs()} yet, then IndexWriter can fail with too many documents. In this case, we have to fail
+     * the engine because we already generated sequence numbers for write operations; otherwise we will have gaps in sequence numbers.
+     * To avoid this, we keep track the number of documents that are being added to IndexWriter, and account it in
+     * {@link InternalEngine#tryReserveDocs(Operation, int)}. Although we can double count some adding documents in both IW and Engine,
+     * this shouldn't be an issue because it happens for a short window and we adjust the reservingNumDocs once an indexing is completed.
+     */
+    private final AtomicLong reservingNumDocs = new AtomicLong();
```

Comment:
Naming nit: `reserved` feels rather too general a term for me, how would you feel about renaming this something like `inFlightDocCount` instead?

Reply:
++